hydra-node/src/Hydra/Events/SqlLiteBased.hs

```diff
@@ -0,0 +1,48 @@
+-- | A sqllite-based event source and sink using JSON encoding.
+--
+-- This serves as an example of how to create an 'EventSource' and 'EventSink'.
+module Hydra.Events.SqlLiteBased where
+
+import Hydra.Prelude
+
+import Conduit (mapMC, (.|))
+import Control.Concurrent.Class.MonadSTM (newTVarIO, writeTVar)
+import Hydra.Events (EventSink (..), EventSource (..), HasEventId (..))
+import Hydra.SqlLitePersistence (PersistenceIncremental (..))
+
+-- | A basic sqllite based event source and sink defined using an
+-- 'PersistenceIncremental' handle.
+eventPairFromPersistenceIncremental ::
+  (ToJSON e, FromJSON e, HasEventId e, MonadSTM m) =>
+  PersistenceIncremental e m ->
+  m (EventSource e m, EventSink e m)
+eventPairFromPersistenceIncremental PersistenceIncremental{append, source} = do
+  eventIdV <- newTVarIO Nothing
+  let
+    getLastSeenEventId = readTVar eventIdV
+
+    setLastSeenEventId evt = do
+      writeTVar eventIdV (Just $ getEventId evt)
+
+    -- Keep track of the last seen event id when loading
+    sourceEvents =
+      source
+        .| mapMC
+          ( \event -> lift . atomically $ do
+              setLastSeenEventId event
+              pure event
+          )
+
+    -- Filter events that are already stored
+    putEvent evt = do
+      atomically getLastSeenEventId >>= \case
+        Nothing -> store evt
+        Just lastSeenEventId
+          | getEventId evt > lastSeenEventId -> store evt
```

Comment:
No need to keep track of a last seen event id here as the sqlite database should make it very easy to not write the same key that is already present (or overwrite it).